# Multiple Linear Regression in Statsmodels - Lab

## Introduction
In this lab, you'll practice fitting a multiple linear regression model on our Boston Housing Data set!

## Objectives
You will be able to:
* Run linear regression on Boston Housing dataset with all the predictors
* Interpret the parameters of the multiple linear regression model

## The Boston Housing Data

We pre-processed the Boston Housing Data again. This time, however, we did things slightly different:
- We dropped "ZN" and "NOX" completely
- We categorized "RAD" in 3 bins and "TAX" in 4 bins
- We used min-max-scaling on "B", "CRIM" and "DIS" (and logtransformed all of them first, except "B")
- We used standardization on "AGE", "INDUS", "LSTAT" and "PTRATIO" (and logtransformed all of them first, except for "AGE") 

In [14]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
boston = load_boston()

boston_features = pd.DataFrame(boston.data, columns = boston.feature_names)
boston_features = boston_features.drop(["NOX","ZN"],axis=1)

# first, create bins for based on the values observed. 3 values will result in 2 bins
bins = [0,6,  24]
bins_rad = pd.cut(boston_features['RAD'], bins)
bins_rad = bins_rad.cat.as_unordered()

# first, create bins for based on the values observed. 4 values will result in 3 bins
bins = [0, 270, 360, 712]
bins_tax = pd.cut(boston_features['TAX'], bins)
bins_tax = bins_tax.cat.as_unordered()

tax_dummy = pd.get_dummies(bins_tax, prefix="TAX")
rad_dummy = pd.get_dummies(bins_rad, prefix="RAD")
boston_features = boston_features.drop(["RAD","TAX"], axis=1)
boston_features = pd.concat([boston_features, rad_dummy, tax_dummy], axis=1)

In [15]:
age = boston_features["AGE"]
b = boston_features["B"]
logcrim = np.log(boston_features["CRIM"])
logdis = np.log(boston_features["DIS"])
logindus = np.log(boston_features["INDUS"])
loglstat = np.log(boston_features["LSTAT"])
logptratio = np.log(boston_features["PTRATIO"])

# minmax scaling
boston_features["B"] = (b-min(b))/(max(b)-min(b))
boston_features["CRIM"] = (logcrim-min(logcrim))/(max(logcrim)-min(logcrim))
boston_features["DIS"] = (logdis-min(logdis))/(max(logdis)-min(logdis))

#standardization
boston_features["AGE"] = (age-np.mean(age))/np.sqrt(np.var(age))
boston_features["INDUS"] = (logindus-np.mean(logindus))/np.sqrt(np.var(logindus))
boston_features["LSTAT"] = (loglstat-np.mean(loglstat))/np.sqrt(np.var(loglstat))
boston_features["PTRATIO"] = (logptratio-np.mean(logptratio))/(np.sqrt(np.var(logptratio)))

In [16]:
boston_features.head()

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0  0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1  0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2  0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3  0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4  0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   

      LSTAT  RAD_(0, 6]  RAD_(6, 24]  TAX_(0, 270]  TAX_(270, 360]  \
0 -1.275260           1            0             0               1   
1 -0.263711           1            0             1               0   
2 -1.627858           1            0             1               0   
3 -2.153192           1            0             1               0   
4 -1.162114           1            0             1               0   

   TAX_(360, 712]  
0               0  
1               0  
2               0  
3               0  
4               0

## Run an linear model in Statsmodels

In [6]:
!pip install statsmodels

    100% |████████████████████████████████| 7.4MB 5.2MB/s eta 0:00:01
    100% |████████████████████████████████| 235kB 37.3MB/s ta 0:00:01
You are using pip version 10.0.1, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.


In [7]:
import statsmodels.api as sm

In [17]:
features = boston_features
Y = pd.DataFrame(boston.target, columns= ["price"])

X = features
X_sm = sm.add_constant(X)
model = sm.OLS(y_sm,X_sm).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.779
Model:                            OLS   Adj. R-squared:                  0.774
Method:                 Least Squares   F-statistic:                     144.9
Date:                Mon, 03 Dec 2018   Prob (F-statistic):          5.08e-153
Time:                        18:08:30   Log-Likelihood:                -1458.2
No. Observations:                 506   AIC:                             2942.
Df Residuals:                     493   BIC:                             2997.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const              4.4607      1.789      2.493      0.013       0.946       7.976
CRIM              -1.9538      2.115     -0.924      0.356      -6.110       2.202
INDUS             -0.8046      0.362     -2.220      0.027      -1.517      -0.093
CHAS               2.5959      0.796      3.260      0.001       1.032       4.160
RM                 2.6466      0.408      6.488      0.000       1.845       3.448
AGE                0.0794      0.352      0.226      0.821      -0.612       0.770
DIS              -10.0962      1.856     -5.439      0.000     -13.743      -6.449
PTRATIO           -1.4867      0.241     -6.160      0.000      -1.961      -1.013
B                  3.8412      0.986      3.897      0.000       1.905       5.778
LSTAT             -5.6288      0.354    -15.912      0.000      -6.324      -4.934
RAD_(0, 6]         1.5613      0.821      1.901      0.058      -0.052       3.175
RAD_(6, 24]        2.8993      1.073      2.701      0.007       0.791       5.008
TAX_(0, 270]       2.6222      0.716      3.665      0.000       1.216       4.028
TAX_(270, 360]     1.3624      0.704      1.936      0.053      -0.020       2.745
TAX_(360, 712]     0.4761      0.684      0.696      0.487      -0.868       1.820
==============================================================================
Omnibus:                      106.730   Durbin-Watson:                   1.093
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              432.101
Skew:                           0.891   Prob(JB):                     1.48e-94
Kurtosis:                       7.162   Cond. No.                     5.67e+16
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 6.79e-30. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

## Run the same model in Scikit-learn

In [18]:
from sklearn.linear_model import LinearRegression

In [22]:
linreg = LinearRegression()
linreg.fit(X, y_sm)
print(linreg.intercept_)


[8.17785839]


CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0    0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1    0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2    0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3    0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4    0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   
5    0.162521 -1.778965   0.0  6.430 -0.351157  0.707895  0.165279  0.992990   
6    0.276046 -0.125137   0.0  6.012 -0.070229  0.671500 -1.496563  0.996722   
7    0.327656 -0.125137   0.0  6.172  0.978808  0.700059 -1.496563  1.000000   
8    0.367371 -0.125137   0.0  5.631  1.117494  0.709276 -1.496563  0.974104   
9    0.344658 -0.125137   0.0  6.004  0.616090  0.743201 -1.496563  0.974305   
10   0.373926 -0.125137   0.0  6.377  0.914799  0.727217 -1.496563  0.988956   
11   0.305940 -0.125137   0.0  6.009  0.509409  0.719175 -1.496563  1.000000   
12   0.282362 -0.125137   0.0  5.889 -1.051700  0.663113 -1.496563  0.983862   
13   0.481724 -0.081680   0.0  5.949 -0.240919  0.601338  1.095518  1.000000   
14   0.483078 -0.081680   0.0  6.096  0.566306  0.578763  1.095518  0.957436   
15   0.481329 -0.081680   0.0  5.834 -0.429390  0.582214  1.095518  0.996772   
16   0.535631 -0.081680   0.0  5.935 -1.396638  0.582214  1.095518  0.974658   
17   0.504684 -0.081680   0.0  5.990  0.466736  0.559046  1.095518  0.974406   
18   0.507126 -0.081680   0.0  5.456 -1.137046  0.510723  1.095518  0.727899   
19   0.496582 -0.081680   0.0  5.727  0.032897  0.510723  1.095518  0.984997   
20   0.553642 -0.081680   0.0  5.570  1.049929  0.510878  1.095518  0.948737   
21   0.513370 -0.081680   0.0  5.965  0.733440  0.534015  1.095518  0.988981   
22   0.552013 -0.081680   0.0  6.142  0.822342  0.530282  1.095518  1.000000   
23   0.528914 -0.081680   0.0  5.813  1.117494  0.542632  1.095518  0.994049   
24   0.500052 -0.081680   0.0  5.924  0.907687  0.572839  1.095518  0.993520   
25   0.511947 -0.081680   0.0  5.599  0.608978  0.578073  1.095518  0.764285   
26   0.488506 -0.081680   0.0  5.813  0.772557  0.599050  1.095518  0.949518   
27   0.525396 -0.081680   0.0  6.047  0.719216  0.577960  1.095518  0.771748   
28   0.503177 -0.081680   0.0  6.495  0.918355  0.578083  1.095518  0.977407   
29   0.530388 -0.081680   0.0  6.674  0.665875  0.557172  1.095518  0.957966   
..        ...       ...   ...    ...       ...       ...       ...       ...   
476  0.695890  0.947825   0.0  6.484  0.889907  0.300543  0.784047  0.998260   
477  0.813789  0.947825   0.0  5.304  1.021481  0.261386  0.784047  0.880428   
478  0.773591  0.947825   0.0  6.185  1.000145  0.275157  0.784047  0.956629   
479  0.808870  0.947825   0.0  6.229  0.690768  0.230282  0.784047  0.965757   
480  0.714587  0.947825   0.0  6.242 -0.137794  0.467238  0.784047  1.000000   
481  0.712484  0.947825   0.0  6.750  0.224924  0.455701  0.784047  0.990342   
482  0.712905  0.947825   0.0  7.061  0.299601  0.465562  0.784047  0.995915   
483  0.638603  0.947825   0.0  5.762 -1.005472  0.542950  0.784047  0.989964   
484  0.620842  0.947825   0.0  5.871 -0.948575  0.502600  0.784047  0.934011   
485  0.666348  0.947825   0.0  6.312 -0.592969  0.531847  0.784047  0.979121   
486  0.712182  0.947825   0.0  6.114  0.399171  0.481952  0.784047  0.989359   
487  0.695119  0.947825   0.0  5.905 -0.546740  0.432381  0.784047  0.978113   
488  0.332129  1.497881   0.0  5.454  0.857902  0.201163  0.744248  0.995436   
489  0.352559  1.497881   0.0  5.414  1.057042  0.185729  0.744248  0.866736   
490  0.365481  1.497881   0.0  5.093  1.046373  0.201557  0.744248  0.802133   
491  0.294927  1.497881   0.0  5.983  1.074822  0.211945  0.744248  0.982879   
492  0.300311  1.497881   0.0  5.983  0.530745  0.263227  0.744248  1.000000   
493  0.346652  0.142876   0.0  5.707 -0.518292  0.314279  0.376884  1.000000   
494  0.39

## Remove the necessary variables to make sure the coefficients are the same for Scikit-learn vs Statsmodels

In [ ]:
from

### Statsmodels

### Scikit-learn

## Interpret the coefficients for PTRATIO, PTRATIO, LSTAT

- CRIM: per capita crime rate by town
- INDUS: proportion of non-retail business acres per town
- CHAS: Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
- RM: average number of rooms per dwelling
- AGE: proportion of owner-occupied units built prior to 1940
- DIS: weighted distances to five Boston employment centres
- RAD: index of accessibility to radial highways
- TAX: full-value property-tax rate per $10,000
- PTRATIO: pupil-teacher ratio by town
- B: 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
- LSTAT: % lower status of the population

## Predict the house price given the following characteristics (before manipulation!!)

Make sure to transform your variables as needed!

- CRIM: 0.15
- INDUS: 6.07
- CHAS: 1        
- RM:  6.1
- AGE: 33.2
- DIS: 7.6
- PTRATIO: 17
- B: 383
- LSTAT: 10.87
- RAD: 8
- TAX: 284

## Summary
Congratulations! You've fitted your first multiple linear regression model on the Boston Housing Data.